packages/cli-kit/src/private/node/ui/components/ConcurrentOutput.tsx

```diff
@@ -77,43 +82,84 @@ const ConcurrentOutput: FunctionComponent<ConcurrentOutputProps> = ({
 }) => {
   const [processOutput, setProcessOutput] = useState<Chunk[]>([])
   const {exit: unmountInk} = useApp()
-  const prefixColumnSize = Math.max(...processes.map((process) => process.prefix.length))
-  const concurrentColors: TextProps['color'][] = useMemo(() => ['yellow', 'cyan', 'magenta', 'green', 'blue'], [])
-  const lineColor = useCallback(
-    (index: number) => {
-      const colorIndex = index < concurrentColors.length ? index : index % concurrentColors.length
-      return concurrentColors[colorIndex]!
-    },
-    [concurrentColors],
-  )
-  const writableStream = useCallback(
-    (process: OutputProcess, index: number) => {
-      return new Writable({
-        write(chunk, _encoding, next) {
-          const lines = stripAnsi(chunk.toString('utf8').replace(/(\n)$/, '')).split(/\n/)
-          addOrUpdateConcurrentUIEventOutput({prefix: process.prefix, index, output: lines.join('\n')})
-          setProcessOutput((previousProcessOutput) => [
-            ...previousProcessOutput,
-            {
-              color: lineColor(index),
-              prefix: process.prefix,
-              lines,
-            },
-          ])
-          next()
-        },
-      })
-    },
-    [lineColor],
-  )
+
+  const concurrentColors = [
+    'yellow',
+    '#008080' /*teal*/,
+    'cyan',
+    '#ff8700' /*orange*/,
+    'magenta',
+    '#ffd787' /*gold*/,
+    '#008000', /*green*/
+    '#800080' /*purple*/,
+    '#000080' /*navy*/,
+    '#808000' /*olive*/
+  ]
+  const [prefixColumnSize, setPrefixColumnSize] = useState<number>(12)
+
+  const parseLog = (log: string) : ParsedLog => {
+    // Example: <::hello-world::> foo bar\nssssada
+    const prefixRegex = /(<::(([^:])+)::>\s)[\s\S]+/g
+    const prefixMatch = prefixRegex.exec(log)
+    if (prefixMatch && prefixMatch[1] && prefixMatch[2]) {
+      return {
+        prefix: prefixMatch[2], // hello-world
+        log: log.substring(prefixMatch[1].length) // To strip off <::hello-world::>
+      }
+    }
+    return {
+      log
+    }
+  }
+
+  const addPrefix = (prefix: string, prefixes: string[]) => {
+    const index = prefixes.indexOf(prefix)
+    if (index != -1) {
+      return index
+    }
+    prefixes.push(prefix)
+    setPrefixColumnSize(previousColumnSize => {
+      return Math.max(...prefixes.map((prefix) => prefix.length), previousColumnSize)
+    })
+    return prefixes.length-1
+  }
+
+  const lineColor = (index: number) => {
+    const colorIndex = index % concurrentColors.length
+    return concurrentColors[colorIndex]
+  }
+
+  const writableStream = (process: OutputProcess, prefixes: string[]) => {
+    return new Writable({
+      write(chunk, _encoding, next) {
+        const parsedLog : ParsedLog = parseLog(chunk.toString('utf8'))
+        const prefix = parsedLog.prefix ?? process.prefix
+        const index = addPrefix(prefix, prefixes)
+
+        const lines = stripAnsi(parsedLog.log.replace(/(\n)$/, '')).split(/\n/)
```

Comment:
So we are parsing before splitting lines, do we assume chunk === line?
Could we receive log lines from different functions in the same chunk?

Reply:
🤔 I don't think so? Though we could add support for that later I think by looking for the prefix marker at the start of any line?

Reply:
No longer relevant with the switch to AsyncLocalStorage.